### PR TITLE
(dev) construct service from existing rcl_service_t

### PR DESCRIPTION
**Pull request by Karsten1987**
this PR enables a wrapping around a rcl_service_t which is externally initialized. The goal here is to instantiate a service on the rcl level and still be able to add it into a rclcpp executor. 

On the flipside, I have to have a boolean flag stating whether we own ownership over this service or not. This becomes important when trying to destroy the service.

### CODE REVIEW

**Comment by Karsten1987** [comment]
@dirk-thomas I added some dummy mock messages in order to create a type-support. Can you please verify the `CMakeLists.txt` that these are only used for the test and not being installed?

**Comment by dirk-thomas** [comment]
If you don't want the generated messages to be installed you need to pass the `SKIP_INSTALL` flag (https://github.com/ros2/rosidl/blob/c9545c9d3ea750e288fda40d9cb458e8798a6106/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake#L34-L35).

**Comment by Karsten1987** [comment]
http://ci.ros2.org/job/ci_linux/1947/badge/icon

**Comment by Karsten1987** [comment]
[Build Status](http://ci.ros2.org/job/ci_linux/1948/)

**Comment by Karsten1987** [comment]
@wjwwood any feedback about this?

File: rclcpp/include/rclcpp/service.hpp [hunk 1]
```diff
@@ -66,10 +67,9 @@ class ServiceBase
 protected:
   RCLCPP_DISABLE_COPY(ServiceBase);
 
-  std::shared_ptr<rcl_node_t> node_handle_;
+  rcl_node_t * node_handle_;
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: Why did you change this to a pointer from a shared pointer?

File: rclcpp/src/rclcpp/service.cpp [hunk 2]
```diff
@@ -39,11 +37,11 @@ ServiceBase::~ServiceBase()
 std::string
 ServiceBase::get_service_name()
 {
-  return this->service_name_;
+  return rcl_service_get_service_name(service_handle_);
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/service.cpp]: I understand what you're doing here, but it would be better I think to leave it as is and fix all cases of this (subscription, client, etc.) at the same time. This change is not required to fulfill the main purpose of this pr I think.

Comment (wjwwood) [inline rclcpp/src/rclcpp/service.cpp]: Alternatively, update the other locations where this is happening in this pr. Basically, I don't want an inconsistent state.

File: rclcpp/test/test_externally_defined_services.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    rclcpp, srv, Mock);
+  if (rcl_service_init(&service_handle, node_handle->get_rcl_node_handle(),
+    ts, "base_node_service", &service_options) != RCL_RET_OK)
+  {
+    FAIL();
+    return;
+  }
+  rclcpp::any_service_callback::AnyServiceCallback<rclcpp::srv::Mock> cb;
+  // don't initialize the service
+  // expect fail
```
Comment (wjwwood) [inline rclcpp/test/test_externally_defined_services.cpp]: expect fail? The catch fails here.

File: rclcpp/include/rclcpp/service.hpp [hunk 4]
```diff
@@ -95,30 +95,53 @@ class Service : public ServiceBase
...
 
+  Service(
+    rcl_node_t * node_handle,
+    rcl_service_t * service_handle,
+    AnyServiceCallback<ServiceT> any_callback)
+  : ServiceBase(node_handle), defined_extern(true),
+    any_callback_(any_callback)
+  {
+    // check if service handle was initialized
+    // TODO(Karsten1987): Can this go directly in RCL?
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: https://github.com/ros2/rcl/issues/81

File: rclcpp/include/rclcpp/service.hpp [hunk 5]
```diff
@@ -160,6 +183,8 @@ class Service : public ServiceBase
 private:
   RCLCPP_DISABLE_COPY(Service);
 
+  bool defined_extern = false;
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: Consider a different name, or an unabbreviated name, e.g. `owns_rcl_handle` or `external_rcl_handle`.

File: rclcpp/include/rclcpp/service.hpp (hunk 1 above)
Comment (Karsten1987) [inline rclcpp/include/rclcpp/service.hpp]: The second constructor takes raw `rcl` pointer. I am feeling a bit concerned about turning a raw pointer in a shared pointer since we'd miss 1 ref count. 
But generally, do we need this pointer as a shared component or would a `unique_ptr` help here as the class anyways is marked as non-copyable.

File: rclcpp/test/test_externally_defined_services.cpp (hunk 3 above)
Comment (Karsten1987) [inline rclcpp/test/test_externally_defined_services.cpp]: that must be an erroneous copy-paste

File: rclcpp/include/rclcpp/node.hpp [hunk 6]
```diff
@@ -379,6 +379,11 @@ class Node : public std::enable_shared_from_this<Node>
 
   std::atomic_bool has_executor;
 
+  RCLCPP_PUBLIC
+  void
+  add_service(const rclcpp::service::ServiceBase::SharedPtr & service,
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node.hpp]: If you actually read that post, it points out that copying shared pointers is more expensive compared to copying normal pointers or references, and that using a const reference only becomes a problem if you store the references in a vector or pass it to a function in another thread. Neither of which are happening here. I'd say that const reference is fine here, but you might still choose to not use const reference for other reasons (like the fact that it is a public interface which someone could misuse).

> In controlled circumstances you can pass the shared pointer by constant reference. Be sure that nobody is concurrently deleting the object, though this shouldn't be too hard if you're careful about to whom you give references.

I'd call this a controlled circumstance.

In this past we've solved issues before by removing the pass by reference of shared pointers, but in my opinion that's just using shared pointers to make the functions thread safe and generally doesn't have anything to do with shared pointers. We'd have the same problem if we were using raw pointers, i.e. someone calls this function in one thread and destroys the object (`Node`) in another thread. This is a general thread-safety issue, not a problem with how we're using shared pointers with const reference passing.

File: rclcpp/include/rclcpp/service.hpp (hunk 1 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: Why can the second constructor not take a shared pointer?

**Comment by Karsten1987** [comment]
updated all comments.

**Comment by firesurfer** [comment]
Is this functionality already possible for publisher and subscriptions or are there any special needs for services that make this necessary?

**Comment by wjwwood** [comment]
@firesurfer there is a particular case (@Karsten1987's life cycle work) where we need this ability for Service's in particular. We plan to add support for this to the other classes for symmetry in the future, but we're not doing it here to avoid collisions with other ongoing prs, like my https://github.com/ros2/rclcpp/pull/277.

**Comment by wjwwood** [comment]
Speaking of which, I would really like to get this one merged even if the other life cycle stuff isn't quite ready, so that I can build my pr on top of it.

File: rclcpp/include/rclcpp/service.hpp [hunk 7]
```diff
@@ -100,25 +105,50 @@ class Service : public ServiceBase
...
 
+  Service(
+    std::shared_ptr<rcl_node_t> node_handle,
+    rcl_service_t * service_handle,
+    AnyServiceCallback<ServiceT> any_callback)
+  : ServiceBase(node_handle),
+    any_callback_(any_callback)
+  {
+    // check if service handle was initialized
+    // see: https://github.com/ros2/rcl/issues/81
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: I'd leave the todo, or link here from the that issue. Otherwise we'll never know to come back and fix this.

**Comment by wjwwood** [comment]
Other than a small comment, I'd say merge this as soon as CI is passing for it.

File: rclcpp/CMakeLists.txt [hunk 8]
```diff
@@ -127,6 +127,32 @@ if(BUILD_TESTING)
     )
     target_link_libraries(test_find_weak_nodes ${PROJECT_NAME})
   endif()
+
+  get_default_rmw_implementation(default_rmw)
+  find_package(${default_rmw} REQUIRED)
```
Comment (mikaelarguedas) [inline rclcpp/CMakeLists.txt]: is it a requirement that this test `find_package`s and uses another rmw than all the other tests of this package?

File: rclcpp/include/rclcpp/service.hpp [hunk 9]
```diff
@@ -100,25 +105,50 @@ class Service : public ServiceBase
...
+    if (service_handle->impl == NULL) {
```
Comment (mikaelarguedas) [inline rclcpp/include/rclcpp/service.hpp]: we usually use `NULL == service_handle->impl` to avoid a missing `=` to cause the condition to pass everytime. `if (!service_handle->impl)` would also work

File: rclcpp/include/rclcpp/service.hpp [hunk 10]
```diff
@@ -100,25 +105,50 @@ class Service : public ServiceBase
...
+      throw std::runtime_error(
+              std::string("rcl_service_t in constructor argument ") +
+              "has to be initialized beforehand");
```
Comment (mikaelarguedas) [inline rclcpp/include/rclcpp/service.hpp]: sorry for yet another OCD symptom: This looks a bit weird to me, is that for line length purposes that only half the string is passed to `std::string()`?

Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: This must be a copy-paste issue, normally that pattern is used like this:

```c++
throw std::runtime_error(
  std::string("there was an error while trying to do something: ") +
  rcl_get_error_string_safe());
```

Where the first item needs to be a `std::string` so that the `+` operator works with the `const char *` returned by the error function.

**Comment by mikaelarguedas** [comment]
looks good to me too, all my comments are cosmetic and are not blockers for merging this

**Comment by wjwwood** [comment]
@Karsten1987 can you run CI for this or link some up-to-date successful runs if you already have?

**Comment by Karsten1987** [comment]
Linux: [Build Status](http://ci.ros2.org/job/ci_linux/1957/)

**Comment by Karsten1987** [comment]
Linux: [Build Status](http://ci.ros2.org/job/ci_linux/1957/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/1917/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/1510/)

**Comment by wjwwood** [comment]
Awesome, I'm going to merge this so I can keep hacking on my pr.

**Comment by Karsten1987** [comment]
I'd say this PR is ready to be merged.
